plonky2x/core/src/frontend/eth/rlp/decoder.rs

```diff
@@ -0,0 +1,249 @@
+//! This file implements RLP decoder.
+//!
+//! Reference: https://ethereum.org/en/developers/docs/data-structures-and-encoding/rlp/
+//!
+//! We have a total of 6 different ways of representing Merkle Patricia Tries.
+//! - Encoded versions
+//!     1. Encoded MPT : vec<u8>
+//!     2. Padded Encoded MPT : ([u8; usize], usize)
+//!     3. Circuita Padded Encoded MPT : ArrayVariable<ByteVariable, ENCODING_LEN>
+//! - Decoded versions
+//!     1'. MPT : RLPItem
+//!     2'. Padded MPT : MPTNodeFixedSize
+//!     3'. Circuit Padded MPT : ArrayVariable<ArrayVariable<ByteVariable, 32>, 17>
+//!
+//! We offer three decoding methods:
+//! - decode: 1 -> 1'
+//! - decode_padded_mpt_node: 2 -> 2'
+//! - DecodeHint: 3 -> 3'
+//!
+//! Type converstions can be done as following:
+//! - Converstions between encoded types
+//!     - 1 -> 2: Using the to_fixed_size method
+//!     - 2 -> 3: TODO?
+//! - Converstions between decoded types
+//!     - 3' -> 2': TODO?
+//!     - 2' -> 1': a.iter().take(len).clone().collect()
+//!
+//! 1 and 1' are in this file, decoder.rs. 2 and 2' are in utils.rs. Finally, 3 and 3' are in
+//! builder.rs.
+
+/// An item is a string (i.e., byte array) or a list of items.
+#[derive(PartialEq, Debug)]
+pub enum RLPItem {
+    String(Vec<u8>),
+    List(Vec<RLPItem>),
+}
+
+/// Private helper struct to iterate over a vector.
```

Comment:
Why not just use Stream<u8>?

Reply:
I couldn't figure out a way to track the number of bytes we have read from this stream. The information is necessary for the recursive case. When recursively calling this function in a list, RLP only tells us how many _bytes_ we'd need to read. But let me know if there's a way to do so. I don't like having a new private iterator class just for this purpose.

Reply:
oh I completely misunderstood! `Stream` is in our codebase. For whatever reason, I thought it was a native Rust class 🤦  Yes, that makes perfect sense. I'll do that.

Reply:
Fixed!